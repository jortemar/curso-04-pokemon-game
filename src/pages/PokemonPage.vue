<template>
  <h1 v-if="!pokemon"> Espere por favor... </h1>

  <div v-else>
    <h1>¿Quién es este pokémon?</h1>
    <!-- Podemos escribir los atributos de esta forma (pokemon-id
    en lugar de pokemonId) y todo seguiría funcionando igual -->
    <PokemonPicture 
        :pokemon-id="pokemon.id" 
        :show-pokemon="showPokemon" />
    <!-- Aquí recogemos el selection que hemos hecho en PokemonOptions, con el fin de ejecutar 'checkAnswer'
    No es necesario poner paréntesis en la llamada para recoger el parámetro, que está definido en 'checkAnswer' -->
    <PokemonOptions
        :pokemons="pokemonArr"
        @selectionPokemon="checkAnswer" />

    <!-- Necesitamos agrupar h2 y button, pero no necesitamos el div. Si usamos la etiqueta template,
    esta cumplirá la función del div pero no se generará cuando se cree la estructura html -->
    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">
        Nuevo juego
      </button>
    </template>
    
  </div>
</template>

<script>
  import PokemonOptions from '@/components/PokemonOptions'
  import PokemonPicture from '@/components/PokemonPicture'

  import getPokemonOptions from '@/helpers/getPokemonOptions'

  export default {

      components: {PokemonOptions, PokemonPicture},
      // la data es un método que tiene que devolver un objeto
      data() {
        return {
          pokemonArr: [],
          pokemon: null,
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      },
      methods: {
        async mixPokemonArray() {
          // la función asignada es una promesa. Así que el método también puede ser asíncrono
          this.pokemonArr = await getPokemonOptions()

          const rndInt = Math.floor(Math.random() * 4)
          this.pokemon = this.pokemonArr[rndInt]
        },
        checkAnswer(selectedId) {
          
          this.showPokemon = true
          this.showAnswer = true
          
          if (selectedId === this.pokemon.id) {
            this.message = `Correcto, ${this.pokemon.name}`
          } else {
            this.message = `Oops, era ${this.pokemon.name}`
          }
        },
        newGame() {
          this.showPokemon = false
          this.showAnswer = false
          this.pokemonArr = []
          this.pokemon = null
          this.mixPokemonArray()
        }
      },
      mounted() {
        this.mixPokemonArray()      
      }

  }
</script>