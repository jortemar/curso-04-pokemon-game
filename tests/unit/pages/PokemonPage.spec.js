import {shallowMount, mount} from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage Component', () => {
    let wrapper
    let mixPokemonArraySpy

    beforeEach(() => {
        //mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')

        wrapper = shallowMount(PokemonPage)
    })

    test('debe hacer match con el snapshot', () => {

        expect(wrapper.html()).toMatchSnapshot()
    })

    test('debe llamar mixPokemonArray al montar', () => {
        // no funciona porque cuando el evento se lanza al montarse la página, ya es
        // demasiado tarde para espiar el evento. Hay que ponerlo arriba, antes del wrapper
        // const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')

        // Otra opción sería retrasar el montaje del wrapper, invocándolo nuevamente dentro del test

        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        // no necesitamos asignarle variable
        shallowMount(PokemonPage)

        expect(mixPokemonArraySpy).toHaveBeenCalled()

    })

    // podemos evitarnos esos problemas utilizando este test  (SNAPSHOT CON DATA Y SUBS)
    test('debe hacer match con el snapshot cuando cargan los pokemons', () => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()

    })

    // podemos usar mount o shallowMount para montar el wrapper
    // la diferencia es que shallowMount solo simula el montaje
    // de los componentes, mientras que el mount sí es un montaje real
    // y realizará el ciclo de vida de todos los componentes

    test('debe mostrar los componentes de PokemonPictur y PokemonOptions', () => {

        // Pokemon Picture y Pokemon Options deben existir
        // Pokemon Picture attribute pokemonid === 5
        // Pokemon Options attribute pokemons.toBe(true)

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        // La simulación del montaje del componente da lugar a una
        // etiqueta con el nombre del componente, seguido de 'stub'
        expect(wrapper.find('pokemon-picture-stub').exists()).toBeTruthy()
        expect(wrapper.find('pokemon-options-stub').exists()).toBeTruthy()

        expect(wrapper.find(`pokemon-picture-stub`).attributes(`pokemonid`)).toBe('5')
        expect(wrapper.find(`pokemon-options-stub`).attributes(`pokemons`)).toBeTruthy()

    })

    // PROBAR LAS PROPIEDADES REACTIVAS DEL COMPONENTE
    test('pruebas con checkAnswer', async() => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        // con .vm accedemos a los métodos del componente montado
        // console.log(wrapper.vm)

        await wrapper.vm.checkAnswer(5)

        expect(wrapper.find('h2').exists()).toBeTruthy()
        //console.log(wrapper.find('h2').text())
        expect(wrapper.vm.showPokemon).toBe(true)
        // dos formas distintas de hacer la misma comprobación
        expect(wrapper.find('h2').text()).toBe(`Correcto, ${pokemons[0].name}`)
        await wrapper.vm.checkAnswer(10)
        expect(wrapper.vm.message).toBe(`Oops, era ${pokemons[0].name}`)

        // Cuando hacemos una modificación en alguna de las propiedades
        // reactivas, y estamos esperando que esas propiedades realicen
        // alguna modificación en el DOM, tenemos que esperar que se monte
        // la estructura (re-renderización de componentes).
        // Así que habrá que usar 'async' y 'await'

    })




})