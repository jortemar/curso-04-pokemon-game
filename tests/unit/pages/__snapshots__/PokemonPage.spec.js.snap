// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe hacer match con el snapshot 1`] = `<h1> Espere por favor... </h1>`;

exports[`PokemonPage Component debe hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokémon?</h1>
  <!-- Podemos escribir los atributos de esta forma (pokemon-id
    en lugar de pokemonId) y todo seguiría funcionando igual -->
  <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
  <!-- Aquí recogemos el selection que hemos hecho en PokemonOptions, con el fin de ejecutar 'checkAnswer'
    No es necesario poner paréntesis en la llamada para recoger el parámetro, que está definido en 'checkAnswer' -->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!-- Necesitamos agrupar h2 y button, pero no necesitamos el div. Si usamos la etiqueta template,
    esta cumplirá la función del div pero no se generará cuando se cree la estructura html -->
  <!--v-if-->
</div>
`;
